docs/topics/js/js-ir-migration.md

```diff
@@ -0,0 +1,186 @@
+may require adjusting your code. On this page, we've collected known migration issues along with suggested solutions.
+
+Note that this guide can change with time: we're fixing issues and finding new ones. Please help us keep it complete –
+report any issues you face when switching to the IR compiler to our issue tracker [YouTrack](https://kotl.in/issue)
+<!-- or [this Google form](TODO: link) -->.
+
+## Convert JS- and React-related classes and interfaces to external interfaces 
+
+**Issue**: Using Kotlin interfaces and classes (including data classes) that derive from pure JS classes, such as React's `RState` and
+`RProps`, can cause `ClassCastException`. Such exceptions appear because the compiler attempts to work with instances of 
+such classes as Kotlin objects, whereas they actually come from JS.
+
+**Solution**: convert all classes and interfaces that derive from pure JS classes to [external intefaces](js-interop.md#external-interfaces):
+
+```kotlin
+// Replace this
+interface AppState : RState { }
+interface AppProps : RProps { }
+data class CustomComponentState(var name: String) : RState
+```
+
+```kotlin
+// With this
+external interface AppState : RState { }
+external interface AppProps : RProps { }
+external interface CustomComponentState : RState {
+    var name: String
+}
+```
+
+In IntelliJ IDEA, you can use these [Structural search and replace](https://www.jetbrains.com/help/idea/structural-search-and-replace.html)
+templates to automatically mark interfaces as `external`:
+* [Template for `RState`](https://gist.github.com/SebastianAigner/62119536f24597e630acfdbd14001b98)
+* [Template for `RProps`](https://gist.github.com/SebastianAigner/a47a77f5e519fc74185c077ba12624f9)
+
+## Convert properties of external interfaces to var
+
+**Issue**: properties of external interfaces in Kotlin/JS code can't be read-only (`val`) properties because their values can be
+assigned only after the object is created with `jsObject()` or `js()`.
+
+```kotlin
+val myState = js("{}") as CustomComponentState
+myState.name = "name"
+```
+
+**Solution**: convert all properties of external interfaces to `var`:
+
+```kotlin
+// Replace this
+external interface CustomComponentState : RState {
+    val name: String
+}
+```
+
+```kotlin
+// With this
+external interface CustomComponentState : RState {
+    var name: String
+}
+```
+
+## Make boolean properties nullable in external interface
+
+**Issue**: in JavaScript, the `null` or undefined value of a boolean variable equals to `false`. So, boolean properties can be used
+in expressions without being defined. 
+
+```kotlin
+external interface ComponentProps: RProps {
+    var isInitialized: Boolean
+    var visible: Boolean
+}
+```
+
+```kotlin
+val props = js("{}") as ComponentProps
+props.isInitialized = true
+// visible is not initialized - OK in JS – means it's false
+if (props.isInitialized && props.visible) {} // not OK in Kotlin
```

Comment:
Is there another workaround for this that's worth mentioning? Like checking against `undefined`? https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.js/undefined.html

Reply:
@ilgonmic any ideas you could recommend?

Reply:
So, seems that this problem not in checking in if, but when this value is provided to Kotlin overridden method.
So in fact this problem can be reproduced if this value will be provided to button from react-dom
Something like
```
button {
   attrs {
       autoFocus = props.visible // will be class cast exception in depth of this action
   }
}
```